visualization/panda/panda3d_utils.py

```diff
@@ -0,0 +1,239 @@
+            lens.setFilmSize(1, 1)
+        # make aspect ratio looks same as base window
+        aspect_ratio = base.getAspectRatio()
+        lens.setAspectRatio(aspect_ratio)
+        self.cam = base.makeCamera(self.win, scene=self.render, )  # can also be found in base.camList
+        self.cam.reparentTo(self.render)
+        self.cam.setPos(Point3(cam_pos[0], cam_pos[1], cam_pos[2]))
+        self.cam.lookAt(Point3(lookat_pos[0], lookat_pos[1], lookat_pos[2]), Vec3(up[0], up[1], up[2]))
+        self.cam.node().setLens(lens)  # use same len as sys
+        # set up cartoon effect
+        self._separation = 1
+        self.filter = flt.Filter(self.win, self.cam)
+        self.filter.setCartoonInk(separation=self._separation)
+
+        # camera in camera 2d
+        self.cam2d = base.makeCamera2d(self.win, )
+        self.cam2d.reparentTo(self.render2d)
+        # attach GPTop to the render2d to make sure the DirectGui can be used
+        self.aspect2d = self.render2d.attachNewNode(PGTop("aspect2d"))
+        # self.aspect2d.setScale(1.0 / aspect_ratio, 1.0, 1.0)
+
+        # setup mouse for the new window
+        # name of mouse watcher is to adapt to the name in the input manager
+        self.mouse_thrower = base.setupMouse(self.win, fMultiWin=True)
+        self.mouseWatcher = self.mouse_thrower.getParent()
+        self.mouseWatcherNode = self.mouseWatcher.node()
+        self.aspect2d.node().setMouseWatcher(self.mouseWatcherNode)
+        # self.mouseWatcherNode.addRegion(PGMouseWatcherBackground())
+
+        # setup input manager
+        self.inputmgr = im.InputManager(self, lookatpos=lookat_pos)
+
+        # copy attributes and functions from base
+        ## change the bound function to a function, and bind to `self` to become a unbound function
+        self._interaction_update = functools.partial(base._interaction_update.__func__, self)
+        self.p3dh = base.p3dh
+
+        base.taskMgr.add(self._interaction_update, "interaction_extra_window", appendTask=True)
+
+    def getAspectRatio(self):
+        return self._base.getAspectRatio(self.win)
+
+    def set_win_props(self,
+                      title: str,
+                      size: tuple):
+        """
+        set properties of extra window
+        :param title: the title of the window
+        :param size: 1x2 tuple describe width and height
+        :return:
+        """
+        win_props = WindowProperties()
+        win_props.setSize(size[0], size[1])
+        win_props.setTitle(title)
+        self.win.requestProperties(win_props)
+
+    def set_origin(self, origin: np.ndarray):
+        """
+        :param origin: 1x2 np array describe the left top corner of the window
+        """
+        win_props = WindowProperties()
+        win_props.setOrigin(origin[0], origin[1])
+        self.win.requestProperties(win_props)
+
+    @property
+    def size(self):
+        size = self.win.getProperties().size
+        return np.array([size[0], size[1]])
```

Comment:
property is preferred to be stated early in a class definition

Reply:
Updated